### PR TITLE
Declarative style for particle update comm

### PR DESCRIPTION
**Pull request by fweik**
and reduce duplication.

### CODE REVIEW

**Comment by fweik** [comment]
@KaiSzuttor the AMD issue is unrelated.

**Comment by mkuron** [comment]
AMD fixed in #2412.

**Comment by RudolfWeeber** [comment]
Looks good. A few doc comments would be helpful, though.

```diff
diff --git a/src/core/particle_data.cpp b/src/core/particle_data.cpp
+template <typename S, S Particle::*s, typename T, T S::*m>
+struct UpdateParticle {
+  int id;
+  T value;
+
+  void operator()(Particle &p) const { (p.*s).*m = value; }
+
+  template <class Archive> void serialize(Archive &ar, long int) {
+    ar &id;
+    ar &value;
+  }
+};

+
+template <typename T, T ParticleProperties::*m>
+using UpdateProperty = UpdateParticle<ParticleProperties, &Particle::p, T, m>;
+template <typename T, T ParticlePosition ::*m>
+using UpdatePosition = UpdateParticle<ParticlePosition, &Particle::r, T, m>;
+template <typename T, T ParticleMomentum ::*m>
+using UpdateMomentum = UpdateParticle<ParticleMomentum, &Particle::m, T, m>;
+template <typename T, T ParticleForce ::*m>
+using UpdateForce = UpdateParticle<ParticleForce, &Particle::f, T, m>;
```
Please insert some docs, also with regards to the template params
```diff
+template <typename S, S Particle::*s, typename T, T S::*m> struct message_type;
```
Please also insert some comments here about how the message type distinciotn works.
```diff
+

+template <typename S, S Particle::*s, typename T, T S::*m>
+using message_type_t = typename message_type<S, s, T, m>::type;
+
+struct UpdateParticleVisitor : public boost::static_visitor<void> {
+  template <typename Message> void operator()(const Message &msg) const {
+    assert(local_particles[msg.id]);
+    msg(*local_particles[msg.id]);
+  }
+};
```
Please mention here via what mechanism the actual update gets called since not all readers of this might be familiar with the pattern.

**Comment by KaiSzuttor** [comment]
collision detection seems to have issues

**Comment by fweik** [comment]
Yeah, I called the master instead of the local interface...

**Comment by KaiSzuttor** [comment]
please merge the current python branch

**Comment by fweik** [comment]
@KaiSzuttor

File: src/python/espressomd/particle_data.pyx [hunk 1]
```diff
@@ -368,8 +366,7 @@ cdef class ParticleHandle(object):
             IF MASS == 1:
                 check_type_or_throw_except(
                     _mass, 1, float, "Mass has to be 1 floats")
-                if set_particle_mass(self._id, _mass) == 1:
-                    raise Exception("Set particle position first.")
+                set_particle_mass(self._id, _mass)
```
Comment (KaiSzuttor) [inline src/python/espressomd/particle_data.pyx:365]: why did you remove the exception here?

Comment (fweik) [reply src/python/espressomd/particle_data.pyx:365]: Because I removed the return value from the setter, and forgot it for all the other particles.
I *think* this can only fail if the particle does not actually exist (module some internal errors
that should be handled by asserts), in which case I would throw from the core. The error
message is also misleading.

Comment (fweik) [reply src/python/espressomd/particle_data.pyx:365]: So I guess I'm going to remove the error codes from all the functions? Almost all of them have a hard-coded return value anyhow.

**Comment by fweik** [comment]
@KaiSzuttor I removed the unused return values and cleaned up a bit.

**Review by KaiSzuttor — APPROVED** [review]
